docs/source/hsm.md

```diff
@@ -0,0 +1,161 @@
+endorse transactions without exposing their private keys. Currently, Fabric only
+supports the PKCS11 standard to communicate with an HSM.
+
+## Configuring an HSM
+
+To use an HSM with your Fabric node, you need to update the BCCSP (Crypto Service
+Provider) section of the node configuration file such as core.yaml or
+orderer.yaml. In BCCSP section, you need to select PKCS11 as the provider and
+provide the path to the PKCS11 library that you would like to use. You also need
+to provide the label and pin of the token that you created for your cryptographic
+operations. You can use one token to generate and store multiple keys.
+
+The prebuilt Hyperledger Fabric Docker images are not enabled to use PKCS11. If
+you are deploying Fabric using docker, you need to build your own images and
+enable PKCS11 using the following command:
+```
+make docker GO_TAGS=pkcs11
+```
+You also need to ensure that the PKCS11 library is available to be used by the
+node by installing it or mounting it inside the container.
+
+### Example
+
+The following example demonstrates how to configure a Fabric node to use an HSM.
+
+First, you will need to install an implementation of the PKCS11 interface. This
+example uses the [softhsm](https://github.com/opendnssec/SoftHSMv2) open source
+implementation. After downloading and configuring softhsm, you will need to set
+the SOFTHSM2_CONF environment variable to point to the softhsm2 configuration
+file.
+
+You can then use softhsm to create the token that will handle the cryptographic
+operations of your Fabric node inside an HSM slot. In this example, we create a
+token labelled "fabric" and set the pin to "71811222". After you have created
+the token, update the configuration file to use PKCS11 and your token as the
+crypto service provider. You can find an example BCCSP section below:
+
+```
+#############################################################################
+# BCCSP (BlockChain Crypto Service Provider) section is used to select which
+# crypto library implementation to use
+#############################################################################
+bccsp:
+  default: PKCS11
+  pkcs11:
+    Library: /etc/hyperledger/fabric/libsofthsm2.so
+    Pin: 71811222
+    Label: fabric
+    hash: SHA2
+    security: 256
+    Immutable: false
+```
+
+By default, when private keys are generated using the HSM, the private key is mutable, meaning PKCS11 private key  attributes can be changed after the key is generated. Setting `Immutable` to `true` means that the private key attributes cannot be altered after key generation. Before you configure immutability by setting `Immutable: true`, ensure that PKCS11 object cloning is supported by the HSM.
+
+If you are using AWS HSM there is an additional step required:
+
+- Add the parameter, `AltId` to the `pkcs11` section of the `bccsp` block. Create a long secure string outside of Fabric and assign it to the `AltId` parameter. For example:
+
```

Comment:
Can we mention why AltId is required and what it is used for?

Reply:
Here is flow : 

1. Key pair is generated using PKCS11 interface.
2. SKI(Subject Key Identifier) is calculated based on KeyPair generated in step 1.
3. SKI value is set to PKCS11 CKA_ID field attribute of Public and Private key.

Step 3 is not supported in AWS. So , once crypto object is generated , you cannot modify it.

To circumvent problem with step 3 and for minimal code impact , a new variable was created named AltId. (The change was put by luc). AltId is independent of Private key , it is being generated before key pair is generated. AltId is being set for CKA_ID field of Private Key. @denyeart , please let me know , if you have more comments.

Reply:
Makes sense, please help Pam put that into a couple doc sentences.

Reply:
Add the parameter, `AltId` to the `pkcs11` section of the `bccsp` block. When AWS HSM is being used, this parameter is used to assign a unique value to the CKA_Id attribute of the private and public key.